features/cellular/framework/targets/QUECTEL/BC95/QUECTEL_BC95_CellularStack.cpp

```diff
@@ -212,6 +231,30 @@ nsapi_size_or_error_t QUECTEL_BC95_CellularStack::socket_sendto_impl(CellularSoc
         return sent_len;
     }
 
+    // check for network congestion
+    device_err_t err = _at.get_last_device_error();
+    if (err.errType == DeviceErrorTypeErrorCME && (err.errCode == AT_UART_BUFFER_ERROR || err.errCode == AT_BACK_OFF_TIMER)) {
+        if (socket->proto == NSAPI_UDP) {
+            if (retry < 3) {
+                retry++;
+                tr_info("Socket %d sendto EAGAIN", socket->id);
+                rtos::ThisThread::sleep_for(30);
+                _at.clear_error();
+                goto retry_send;
+            }
+            return NSAPI_ERROR_NO_MEMORY;
+        }
+        if (!core_util_atomic_exchange_bool(&socket->txfull_event, true) && !_txfull_event_id) {
```

Comment:
If you need atomic handling of the `txfull_event`, you also will need atomic handling of the `txfull_event_id`.

Indeed, that's the more important one, as the `call_in` has side-effects.

At this point, I'd say keep it simple by protecting both with the `_socket_mutex`. Any reason not to do that? Much as I like atomics, they're a pain to think about compared to mutexes.

Reply:
Atomic implementation is a kind of copy from [ESP8266Interface](https://github.com/ARMmbed/mbed-os/blob/master/components/wifi/esp8266-driver/ESP8266Interface.cpp). I'm not aware of any `call_in` side effects, but if it the queue guarantees that a new event is always generated after `call_in` (also while executing the same event) then additional locking is not needed here. However, if you think mutex is more safe for some reason I can change the code.

Reply:
Discussed offline - short answer is that multiple `call_in`s are fine, it's that this could lose track of what it's queued, so the clean-up attempt in the destructor becomes ineffective.

(Plus the unprotected multithread access to the variable is formally undefined).